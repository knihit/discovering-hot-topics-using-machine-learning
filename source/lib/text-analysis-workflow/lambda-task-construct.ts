/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { Function, FunctionProps } from '@aws-cdk/aws-lambda';
import { Task } from '@aws-cdk/aws-stepfunctions';
import { InvokeFunction } from '@aws-cdk/aws-stepfunctions-tasks';
import { Construct } from '@aws-cdk/core';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';

export interface StepFuncLambdaTaskProps {
    readonly taskName: string,
    readonly lambdaFunctionProps: FunctionProps,
    readonly inputPath?: string | any,
    readonly outputPath?: string | any
}

export class StepFuncLambdaTask extends Construct {

    private executeTask: Task;

    private lambdaFn: Function;

    constructor (scope: Construct, id: string, props: StepFuncLambdaTaskProps) {
        super(scope, id);

        this.lambdaFn = buildLambdaFunction(this, {
            lambdaFunctionProps: props.lambdaFunctionProps,
        });

        this.executeTask = new Task(this, props.taskName, {
            task: new InvokeFunction(this.lambdaFn),
            inputPath: props.inputPath,
            outputPath: props.outputPath
        });
    }

    public get stepFunctionTask(): Task {
        return this.executeTask;
    }

    public get lambdaFunction(): Function {
        return this.lambdaFn;
    }
}