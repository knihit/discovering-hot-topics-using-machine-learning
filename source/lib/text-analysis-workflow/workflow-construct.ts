/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { CfnPolicy, Effect, Policy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Function } from '@aws-cdk/aws-lambda';
import { LogGroup } from '@aws-cdk/aws-logs';
import { CfnStateMachine, Chain, LogLevel, StateMachine, StateMachineType } from '@aws-cdk/aws-stepfunctions';
import { Aws, Construct } from '@aws-cdk/core';
import { LambdaToStepFunction } from '@aws-solutions-constructs/aws-lambda-step-function';
import { DefaultLogGroupProps } from '@aws-solutions-constructs/core';


export interface WorkflowProps {
    readonly stateMachineType?: StateMachineType;
    readonly chain: Chain,
    readonly lambdaFunc?: Function
}

export class Workflow extends Construct {
    private _stMachine: StateMachine

    constructor (scope: Construct, id: string, props: WorkflowProps) {
        super(scope, id);

        if (props.lambdaFunc != null && props.lambdaFunc != undefined) {
            const lambdaStepFunction = new LambdaToStepFunction(this, 'WorkflowEngine', {
                existingLambdaObj: props.lambdaFunc,
                stateMachineProps: {
                    definition: props.chain,
                    ...(props.stateMachineType === StateMachineType.EXPRESS && {
                        stateMachineType: props.stateMachineType
                    })
                }
            });
            this._stMachine = lambdaStepFunction.stateMachine;
            const role = this._stMachine.node.findChild('Role') as Role;
            const cfnDefaultPolicy = role.node.findChild('DefaultPolicy').node.defaultChild as CfnPolicy;

            cfnDefaultPolicy.addMetadata('cfn_nag', {
                rules_to_suppress: [{
                    id: 'W76',
                    reason: 'The policy adds cloudwatch alarms and allows step function to invoke lambda tasks. Suppressing the \
                    SPCM violation as this policy is required to monitor as well as invoke specific tasks'
                }, {
                    id: 'W12',
                    reason: `The 'LogDelivery' actions do not support resource-level authorizations`
                }]
            });
        } else {
            const logGroup = new LogGroup(this, 'statemachine-log-group', DefaultLogGroupProps());
            logGroup.addStream('statemachine-log-stream');

            const _stateMachineRole = new Role(this, 'StateMachineRole', {
                assumedBy: new ServicePrincipal(`states.${Aws.REGION}.amazonaws.com`)
            });

            const stateMachineLogPolicy = new Policy(this, 'StateMachineLogPolicy');
            stateMachineLogPolicy.addStatements(new PolicyStatement({
                actions: [
                    "logs:CreateLogDelivery",
                    "logs:GetLogDelivery",
                    "logs:UpdateLogDelivery",
                    "logs:DeleteLogDelivery",
                    "logs:ListLogDeliveries",
                    "logs:PutResourcePolicy",
                    "logs:DescribeResourcePolicies",
                    "logs:DescribeLogGroups"
                ],
                resources: [ '*' ],
                effect: Effect.ALLOW
            }));

            stateMachineLogPolicy.attachToRole(_stateMachineRole);
            (stateMachineLogPolicy.node.defaultChild as CfnPolicy).cfnOptions.metadata = {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W12',
                        reason: 'The stepfunction log policy requires that resources be "*"'
                    }]
                }
            }
            this._stMachine = new StateMachine(this, 'WorkflowEngine', {
                definition: props.chain,
                ...(props.stateMachineType === StateMachineType.EXPRESS && {
                    stateMachineType: props.stateMachineType,
                    role: _stateMachineRole!
                })
            });

            const cfnStateMachine = this._stMachine.node.defaultChild as CfnStateMachine;
            if (props.stateMachineType === StateMachineType.EXPRESS) {
                cfnStateMachine.addPropertyOverride(
                    'LoggingConfiguration', {
                        Destinations: [{
                            CloudWatchLogsLogGroup: {
                                LogGroupArn: logGroup!.logGroupArn
                            }
                        }],
                        IncludeExecutionData: true,
                        Level: LogLevel.ALL
                    }
                );
            }
        }
    }

    public get stateMachine(): StateMachine {
        return this._stMachine;
    }
}