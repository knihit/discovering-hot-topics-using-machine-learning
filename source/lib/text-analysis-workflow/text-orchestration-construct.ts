/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { EventBus } from '@aws-cdk/aws-events';
import { CfnPolicy, Effect, Policy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';
import { Chain, Parallel, Pass, StateMachine, Succeed } from '@aws-cdk/aws-stepfunctions';
import { Construct, Duration } from '@aws-cdk/core';
import { LambdaToS3 } from '@aws-solutions-constructs/aws-lambda-s3';
import { buildS3Bucket } from '@aws-solutions-constructs/core';
import { EventStorage } from '../storage/event-storage-construct';
import { StepFuncLambdaTask } from './lambda-task-construct';
import { Workflow } from './workflow-construct';

export interface TextOrchestrationProps {
    readonly eventBus: EventBus
    readonly textAnalysisNameSpace: string,
    readonly s3LoggingBucket: Bucket,
    readonly lambdaTriggerFunc: Function
}

export class TextOrchestration extends Construct {
    private readonly _stateMachine: StateMachine;
    private eventStorage: EventStorage;
    private _s3Bucket: Bucket;

    constructor (scope: Construct, id: string, props: TextOrchestrationProps) {
        super(scope, id);

        [ this._s3Bucket ] = buildS3Bucket(this, {
            bucketProps: {
                versioned: false,
                serverAccessLogsBucket: props.s3LoggingBucket,
                serverAccessLogsPrefix: `${id}/`
            }
        });

        this.eventStorage = new EventStorage(this, 'RawForTA', {
            compressionFormat: 'UNCOMPRESSED',
            s3Bucket: this._s3Bucket
        });

        // start of embedded text detection
        const lambdaPublishEventPolicy = new Policy(this, 'LambdaEventBusPolicy');
        lambdaPublishEventPolicy.addStatements(new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [ props.eventBus.eventBusArn ],
            actions: [ 'events:PutEvents' ],
        }));

        const imageAnalysisTask = new StepFuncLambdaTask(this, 'ImageAnalysis', {
            taskName: 'AnalyzeImage',
            lambdaFunctionProps: {
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/wf-extract-text-in-image`),
                timeout: Duration.minutes(10)
            }
        });

        const imageBucketLambda = new LambdaToS3(this, 'ImageBucket', {
            existingLambdaObj: imageAnalysisTask.lambdaFunction,
            bucketProps: {
                versioned: false,
                serverAccessLogsBucket: props.s3LoggingBucket,
                serverAccessLogsPrefix: `${id}/`
            }
        });

        imageBucketLambda.s3Bucket!.addToResourcePolicy(new PolicyStatement({
            resources: [ `${imageBucketLambda.s3Bucket!.bucketArn}`, `${imageBucketLambda.s3Bucket!.bucketArn}/*` ],
            actions: [ 's3:List*', 's3:Get*' ],
            principals: [ new ServicePrincipal('rekognition.amazonaws.com') ],
            effect: Effect.ALLOW
        }));

        const detectTextLambdaPolicy = new Policy(this, 'TextRekAnalyze', {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ '*' ],
                    actions: [ 'rekognition:detectText' ]
                })
            ]
        });
        const analyzeImageLambdaRole = imageBucketLambda.lambdaFunction.role as Role
        detectTextLambdaPolicy.attachToRole(analyzeImageLambdaRole);

        (detectTextLambdaPolicy.node.defaultChild as CfnPolicy).cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                  id: 'W12',
                  reason: `The * resource allows lambda function to access Amazon Rekognition services. The Rekognition services do not have a resource arn.
                  This permission is retricted to the lambda function responsible for accessing the Amazon Rekognition service`
                }]
            }
        };
        // end of embedded text

        // start of content moderation
        const moderationLabelTask = new StepFuncLambdaTask(this, 'ModerationLabels', {
            taskName: 'DetectModerationLabels',
            lambdaFunctionProps: {
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/wf-detect-moderation-labels`),
                timeout: Duration.minutes(5),
                environment: {
                    S3_BUCKET_NAME: imageBucketLambda.s3Bucket!.bucketName
                }
            }
        });
        imageBucketLambda.s3Bucket!.grantReadWrite(moderationLabelTask.lambdaFunction);

        const detectModerationLambdaPolicy = new Policy(this, 'LabelsRekAnalyze', {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ '*' ],
                    actions: [ 'rekognition:detectModerationLabels' ]
                })
            ]
        });

        const moderationLabelRole = moderationLabelTask.lambdaFunction.role as Role;
        detectModerationLambdaPolicy.attachToRole(moderationLabelRole);

        (detectModerationLambdaPolicy.node.defaultChild as CfnPolicy).cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                  id: 'W12',
                  reason: `The * resource allows lambda function to access Amazon Rekognition services. The Rekognition services do not have a resource arn.
                  This permission is retricted to the lambda function responsible for accessing the Amazon Rekognition service`
                }]
            }
        };
        // end of content moderation

        // start of translation task state lambda function and role policy
        const translateTask = new StepFuncLambdaTask(this, 'Translate', {
            taskName: 'Translate',
            lambdaFunctionProps: {
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/wf-translate-text`),
                timeout: Duration.minutes(5),
                environment: {
                    KINESIS_FIREHOSE_NAME: this.eventStorage.deliveryStreamName
                }
            }
        });

        const lambdaTranslatePolicy = new Policy(this, 'TranslateLambda', {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ `*` ],
                    actions: [ 'translate:translateText' ]
            }), new PolicyStatement({
                effect: Effect.ALLOW,
                resources: [ `${this.eventStorage.kinesisFirehose.attrArn}` ],
                actions: [ "firehose:PutRecord", "firehose:PutRecordBatch" ]
            })]
        });

        const cfnTranslatePolicy = lambdaTranslatePolicy.node.defaultChild as CfnPolicy;
        cfnTranslatePolicy.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                  id: 'W12',
                  reason: `The * resource allows lambda function to access Amazon Translate services. The translate services do not have a resource arn.
                  This permission is retricted to the lambda function responsible for accessing the Amazon Translate service`
                }]
            }
        };

        lambdaTranslatePolicy.attachToRole(translateTask.lambdaFunction.role as Role);
        // end of translation task state lambda function and role policy

        // start of text analysis task state lambda function and role policy
        const textAnalysisTask = new StepFuncLambdaTask(this, 'TextAnalysis', {
            taskName: 'AnalyzeText',
            lambdaFunctionProps: {
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/wf-analyze-text`),
                timeout: Duration.minutes(5)
            }
        });

        const lambdaComprehendPolicy = new Policy(this, 'LambdaComprehendPolicy', {
            statements: [ new PolicyStatement({
                effect: Effect.ALLOW,
                resources: [ `*` ],
                actions: [ 'comprehend:DetectSentiment', 'comprehend:DetectEntities', 'comprehend:DetectKeyPhrases' ]
            })]
        });

        lambdaComprehendPolicy.attachToRole(textAnalysisTask.lambdaFunction.role as Role);

        const cfnPolicy = lambdaComprehendPolicy.node.defaultChild as CfnPolicy;
        cfnPolicy.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                  id: 'W12',
                  reason: `The * resource allows lambda function to access Amazon Comprehend services.
                  The comprehend services not have a resource arn. This permission is retricted to
                   the lambda function responsible for accessing the Amazon Comprehend service`
                }]
            }
        };
        // end of text analysis task state lambda function and role policy

        // start of publish event task state lambda function and role policy
        const publishEventsTask = new StepFuncLambdaTask(this, 'PublishEvents', {
            taskName: 'PublishEvent',
            lambdaFunctionProps: {
                runtime: Runtime.NODEJS_12_X,
                handler: 'index.handler',
                code: Code.fromAsset(`${__dirname}/../../lambda/wf-publish-text-inference`),
                timeout: Duration.minutes(5),
                environment: {
                    EVENT_BUS_NAME: props.eventBus.eventBusName,
                    EVENT_NAMESPACE: props.textAnalysisNameSpace
                },
                memorySize: 256
            }
        });

        lambdaPublishEventPolicy.attachToRole(publishEventsTask.lambdaFunction.role as Role);
        // end of publish event task state lambda function and role policy

        // start of creating workflow
        const parallel = new Parallel(this, 'TextImageSplitProcess', {
            comment: 'Parallely process Text and Image',
            inputPath: '$',
            outputPath: '$'
        });

        const textAnalysisChain = Chain.start(translateTask.stepFunctionTask)
                                        .next(textAnalysisTask.stepFunctionTask)

        parallel.branch(moderationLabelTask.stepFunctionTask,textAnalysisChain);

        const workflowChain = Chain.start(imageAnalysisTask.stepFunctionTask)
                             .next(parallel)
                             .next(new Pass(this, 'MergeJson', {
                                parameters: {
                                    'account_name.$': '$[1].account_name',
                                    'platform.$': '$[1].platform',
                                    'search_query.$': '$[1].search_query',
                                    'feed.$': '$[1].feed',
                                    'Sentiment.$': '$[1].Sentiment',
                                    'SentimentScore.$': '$[1].SentimentScore',
                                    'KeyPhrases.$': '$[1].KeyPhrases',
                                    'Entities.$': '$[1].Entities',
                                    'moderation_labels_in_imgs.$' : '$[0].moderation_labels_in_imgs',
                                    'text_in_images.$': '$[1].text_in_images'
                                }
                             }))
                             .next(publishEventsTask.stepFunctionTask)
                             .next(new Succeed(this, 'Success'));
        // end of creating workflow

        this._stateMachine = new Workflow(this, 'StateMachine', {
            chain: workflowChain,
            lambdaFunc: props.lambdaTriggerFunc
        }).stateMachine;

        // const stateMachinePolicy = this._stateMachine.node.findChild('StateMachine').node.findChild('WorkflowEngine').node.findChild('StateMachine').node.findChild('Role').node.findChild('DefaultPolicy') as CfnPolicy;
        // stateMachinePolicy.cfnOptions.metadata = {
        //     cfn_nag: {
        //         rules_to_suppress: [{
        //           id: 'W76',
        //           reason: `The statemachine policy is to allow different lambda function tasks to be invoked from the state machine`
        //         }]
        //     }
        // };
    }

    public get stateMachine(): StateMachine {
        return this._stateMachine;
    }

    public get s3Bucket(): Bucket {
        return this._s3Bucket;
    }
}