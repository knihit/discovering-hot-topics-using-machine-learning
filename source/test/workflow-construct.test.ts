/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Chain, StateMachineType, Succeed, Wait, WaitTime } from '@aws-cdk/aws-stepfunctions';
import * as cdk from '@aws-cdk/core';
import { Duration } from '@aws-cdk/core';
import { Workflow } from '../lib/text-analysis-workflow/workflow-construct';


test('test standard workflow stack', () => {
    const stack = new cdk.Stack();
    new Workflow (stack, 'WorkflowStack', {
        chain: Chain.start(new Wait(stack, 'WaitState', {
                        time: WaitTime.duration(Duration.minutes(1)),
                    }))
                    .next(new Succeed(stack, 'Success'))
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('test express workflow stack', () => {
    const stack = new cdk.Stack();
    new Workflow (stack, 'WorkflowStack', {
        stateMachineType: StateMachineType.EXPRESS,
        chain: Chain.start(new Wait(stack, 'WaitState', {
                        time: WaitTime.duration(Duration.minutes(1)),
                    }))
                    .next(new Succeed(stack, 'Success'))
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});