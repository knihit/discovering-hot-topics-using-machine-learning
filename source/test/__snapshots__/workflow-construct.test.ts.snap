// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test express workflow stack 1`] = `
Object {
  "Resources": Object {
    "WorkflowStackStateMachineLogPolicyB222B557": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The stepfunction log policy requires that resources be \\"*\\"",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkflowStackStateMachineLogPolicyB222B557",
        "Roles": Array [
          Object {
            "Ref": "WorkflowStackStateMachineRole56AD52FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkflowStackStateMachineRole56AD52FB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkflowStackWorkflowEngine48A8BD9B": Object {
      "DependsOn": Array [
        "WorkflowStackStateMachineRole56AD52FB",
      ],
      "Properties": Object {
        "DefinitionString": "{\\"StartAt\\":\\"WaitState\\",\\"States\\":{\\"WaitState\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":60,\\"Next\\":\\"Success\\"},\\"Success\\":{\\"Type\\":\\"Succeed\\"}}}",
        "LoggingConfiguration": Object {
          "Destinations": Array [
            Object {
              "CloudWatchLogsLogGroup": Object {
                "LogGroupArn": Object {
                  "Fn::GetAtt": Array [
                    "WorkflowStackstatemachineloggroup8D8C9535",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": true,
          "Level": "ALL",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "WorkflowStackStateMachineRole56AD52FB",
            "Arn",
          ],
        },
        "StateMachineType": "EXPRESS",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "WorkflowStackstatemachineloggroup8D8C9535": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "WorkflowStackstatemachineloggroupstatemachinelogstream14FC952A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "WorkflowStackstatemachineloggroup8D8C9535",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`test standard workflow stack 1`] = `
Object {
  "Resources": Object {
    "WorkflowStackStateMachineLogPolicyB222B557": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The stepfunction log policy requires that resources be \\"*\\"",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WorkflowStackStateMachineLogPolicyB222B557",
        "Roles": Array [
          Object {
            "Ref": "WorkflowStackStateMachineRole56AD52FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkflowStackStateMachineRole56AD52FB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkflowStackWorkflowEngine48A8BD9B": Object {
      "DependsOn": Array [
        "WorkflowStackWorkflowEngineRoleF56F2281",
      ],
      "Properties": Object {
        "DefinitionString": "{\\"StartAt\\":\\"WaitState\\",\\"States\\":{\\"WaitState\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":60,\\"Next\\":\\"Success\\"},\\"Success\\":{\\"Type\\":\\"Succeed\\"}}}",
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "WorkflowStackWorkflowEngineRoleF56F2281",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "WorkflowStackWorkflowEngineRoleF56F2281": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkflowStackstatemachineloggroup8D8C9535": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "WorkflowStackstatemachineloggroupstatemachinelogstream14FC952A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "WorkflowStackstatemachineloggroup8D8C9535",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
